File: app/src/main/java/com/abs/launcher/LauncherProvider.java
package com.abs.launcher;

import android.content.ContentProvider;
import android.content.ContentValues;
import android.content.Context;
import android.content.UriMatcher;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.net.Uri;

import com.abs.launcher.util.Utils;

import java.util.List;

/**
 * Created by zy on 17-5-18.
 */

public class LauncherProvider extends ContentProvider {

    public static final String AUTHORITY = "com.abs.launcher.Provider";

    private static final String DATABASE_NAME = "launcher.db";
    private static final int DATABASE_VERSION = 1;

    private SQLiteOpenHelper mDbHelper;

    private static final int MATCH_CODE_ITEM = 1;
    private static final int MATCH_CODE_DIR = 2;
    private static UriMatcher URI_MATCHER = new UriMatcher(UriMatcher.NO_MATCH);
    static {
        URI_MATCHER.addURI(AUTHORITY, "*/#", MATCH_CODE_ITEM);
        URI_MATCHER.addURI(AUTHORITY, "*", MATCH_CODE_DIR);
    }

    @Override
    public boolean onCreate() {
        mDbHelper = new DataBaseHelper(this.getContext());
        return true;
    }

    @Override
    public Cursor query(Uri uri, String[] projection, String selection, String[] selectionArgs, String sortOrder) {
        SqlArguments sql = new SqlArguments(uri);
        SQLiteDatabase db = mDbHelper.getReadableDatabase();
        return db.query(sql.table, projection, selection, selectionArgs, null, null, sortOrder);
    }

    @Override
    public String getType(Uri uri) {
        SqlArguments sql = new SqlArguments(uri);
        switch (URI_MATCHER.match(uri)) {
            case MATCH_CODE_ITEM:
                return "vnd.android.cursor.item/vnd.launcher." + sql.table;
            case MATCH_CODE_DIR:
                return "vnd.android.cursor.dir/vnd.launcher." + sql.table;
        }
        return null;
    }

    @Override
    public Uri insert(Uri uri, ContentValues values) {
        SqlArguments sql = new SqlArguments(uri);
        SQLiteDatabase db = mDbHelper.getWritableDatabase();
        long id = db.insert(sql.table, null, values);
        return Uri.parse("content://" + AUTHORITY + "/" + sql.table + "/" + id);
    }

    @Override
    public int delete(Uri uri, String selection, String[] selectionArgs) {
        SqlArguments sql = new SqlArguments(uri);
        SQLiteDatabase db = mDbHelper.getWritableDatabase();
        return db.delete(sql.table, selection, selectionArgs);
    }

    @Override
    public int update(Uri uri, ContentValues values, String selection, String[] selectionArgs) {
        SqlArguments sql = new SqlArguments(uri);
        SQLiteDatabase db = mDbHelper.getWritableDatabase();
        return db.update(sql.table, values, selection, selectionArgs);
    }

    private class DataBaseHelper extends SQLiteOpenHelper {

        public DataBaseHelper(Context context) {
            super(context, DATABASE_NAME, null, DATABASE_VERSION);
        }

        @Override
        public void onCreate(SQLiteDatabase db) {
            db.execSQL("CREATE TABLE IF NOT EXISTS " + DbManager.TABLE_FAVORITE + "(" +
                    DbManager.generateSqlCreation(DbManager.Favorites.class) +
                    ")");
            db.execSQL("CREATE TABLE IF NOT EXISTS " + DbManager.TABLE_ALL_APPS + "(" +
                    DbManager.generateSqlCreation(DbManager.Applications.class) +
                    ")");
            db.execSQL("CREATE TABLE IF NOT EXISTS " + DbManager.TABLE_HIDDEN_APPS + "(" +
                    DbManager.generateSqlCreation(DbManager.HiddenApplications.class) +
                    ")");
        }

        @Override
        public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {

        }
    }

    private static class SqlArguments {
        String table;
        String id;
        boolean notify;

        SqlArguments(Uri uri) {
            notify = Boolean.parseBoolean(uri.getQueryParameter("notify"));
            List<String> paths = uri.getPathSegments();
            if (!Utils.isEmpty(paths)) {
                switch (paths.size()) {
                    case 2:
                        id = paths.get(1);
                    case 1:
                        table = paths.get(0);
                        return;
                }
            }
            throw new IllegalArgumentException("invalid uri");
        }
    }
}

File: app/src/main/java/com/abs/launcher/LauncherProvider.kt
package com.abs.launcher

import android.content.ContentProvider
import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.net.Uri
import com.abs.launcher.model.db.Applications
import com.abs.launcher.model.db.Favorites
import com.abs.launcher.model.db.generateSQL

/**
 * Created by zy on 18-1-7.
 */
class LauncherProvider : ContentProvider() {

    private lateinit var database: DatabaseHelper
    private val BASE_URI = Uri.parse("content://$AUTHORITY")

    override fun onCreate(): Boolean {
        database = DatabaseHelper(context)
        return true
    }

    override fun insert(uri: Uri, values: ContentValues): Uri {
        val sql = SqlArgument(uri)
        val db = database.writableDatabase
        val id = db.insert(sql.table, null, values)
        if (id.toInt() == -1) {
            return BASE_URI
        }
        return uri.buildUpon().appendPath(id.toString()).build()
    }

    override fun query(uri: Uri, projection: Array<out String>?, selection: String?, selectionArgs: Array<out String>?, sortOrder: String?): Cursor {
        val sql = SqlArgument(uri)
        val db = database.readableDatabase
        return db.query(sql.table, projection, selection, selectionArgs, null, null, sortOrder)
    }

    override fun update(uri: Uri, values: ContentValues?, selection: String?, selectionArgs: Array<out String>?): Int {
        val sql = SqlArgument(uri)
        val db = database.writableDatabase
        return db.update(sql.table, values, selection, selectionArgs)
    }

    override fun delete(uri: Uri, selection: String?, selectionArgs: Array<out String>?): Int {
        val sql = SqlArgument(uri)
        val db = database.writableDatabase
        return db.delete(sql.table, selection, selectionArgs)
    }

    override fun getType(uri: Uri): String {
        val sql = SqlArgument(uri)
        return if (sql.id == null) {
            "vnd.android.cursor.dir/vnd.launcher." + sql.table
        } else {
            "vnd.android.cursor.item/vnd.launcher." + sql.table
        }
    }

    class DatabaseHelper(context: Context) : SQLiteOpenHelper(context, "launcher.db", null, DATABASE_VERSION) {

        override fun onCreate(db: SQLiteDatabase) {
            db.execSQL(generateSQL(Favorites))
            db.execSQL(generateSQL(Applications))
        }

        override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
            db.execSQL("DROP TABLE IF EXISTS ${Favorites.table}")
            db.execSQL("DROP TABLE IF EXISTS ${Applications.table}")
        }

    }

    class SqlArgument(uri: Uri) {
        var table: String? = uri.pathSegments.elementAtOrNull(0)
        var id: String? = uri.pathSegments.elementAtOrNull(1)
        var notify = uri.getQueryParameter("notify")?.toBoolean() ?: false
    }

}